pkg/strutil/k8s.go

```diff
@@ -0,0 +1,31 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package strutil
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+// Pre-compile the regex for efficiency if the function is called often.
+// Kept the original regex for strict matching.
+var podOrdinalRegex = regexp.MustCompile(`^.*\-(\d+)\..+\.svc\.cluster\.local$`)
```

Comment:
Yeah, this is a very strong assumption, can you comment the new hashring algorithm with this strong assumption.